### ISSUE TITLE
Closing wezterm in fullscreen mode kills all wezterm processes with a tiling window manager (dwm)

### ISSUE THREAD

**Issue by huibosa** [wez/wezterm#2155]
### What Operating System(s) are you seeing this problem on?

Linux X11

### WezTerm version

20220612-201748-75066cb5

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

I'm using dwm with two monitors:

1. In `MONITOR1`, open wezterm and *maximize* it (maximization provided by window manager, I've set `<Alt-Shift-f>` to it).
2. Switch to `MONITOR2`, open wezterm.
3. Close wezterm in `MONITOR2`, there is a chance (about 20% maybe) that **all** wezterms (both in `MONITOR1` and `MONITOR2`, and even previously opened in different workspaces) will be closed.
4. All wezterm processes is killed

### Configuration

no config

**Comment by wez** [comment]
Sounds like your window manager is killing the wezterm gui process via `XKillClient` rather than using the `WM_DELETE` protocol to close the individual window.

Can you verify that `xprop WM_PROTOCOLS` shows something like this when used on the wezterm windows:

```
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
```

**Comment by huibosa** [comment]
> Sounds like your window manager is killing the wezterm gui process via `XKillClient` rather than using the `WM_DELETE_WINDOW` protocol to close the individual window.
> 
> Can you verify that `xprop WM_PROTOCOLS` shows something like this when used on the wezterm windows:
> 
> ```
> WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
> ```

The output is:

```
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
```

**Comment by wez** [comment]
Then the question is: why is the WM killing the client instead of using the delete window protocol?
When this issue triggers, are you quickly opening and closing the windows?
I wonder if perhaps there is a race condition where the property hasn't been set on the window yet if you are quickly opening and closing them.

**Comment by huibosa** [comment]
> Then the question is: why is the WM killing the client instead of using the delete window protocol? When this issue triggers, are you quickly opening and closing the windows? I wonder if perhaps there is a race condition where the property hasn't been set on the window yet if you are quickly opening and closing them.

* When I quickly open and close window, such as open massive of wezterms in one window and the press `<Alt-Shift-q>` (my key binding for closing window) for one second or more, all process will be killed, with a probability of occurrence of nearly 100%.
* When not closing quickly, the phenomenon also happens, but with lower probability

**Comment by huibosa** [comment]
> I've pushed a commit that flushes pending commands to the X server prior to mapping the window. Maybe it will help?
> 
> It should be available in the nightly builds produced by CI within about 20 minutes.
> 
> If you are eager and can [build from source](https://wezfurlong.org/wezterm/install/source.html) then you may be able to try this out more quickly.

I've tried the latest build

```
wezterm 20220622-113325-21f1e632
```

But the bug is still exist and can be steadily reproduce by the above mentioned method.

**Comment by huibosa** [comment]
> Honestly, this sounds like a dwm bug to me.
> 
> Closing windows works correctly in other window managers.

Maybe you are right, anyway, thanks for your work sir!

**Comment by wez** [comment]
Since this doesn't appear to be a wezterm issue, I'm going to close this!

**Comment by tpoliaw** [comment]
I'm not sure if this is related or whether it is a different bug (happy to open a new issue if you think it is) but I am now getting something very similar. I am using version 20220624-141144-bd1b7c5d on i3wm on X11 and closing a terminal using `ctrl-d` closes all open terminals, killing any running processes in them. It is not every time but I would say more than 50%. If I open two terminals quickly and then hit `ctrl-d` it rises to almost every time.

I am also getting a notification via dbus Notifications/dunst `X(Drawable(ValueError { response_type: 0, error_code: 9, sequence: 17752, bad_value: 6291471, minor_opcode: 2, major_opcode: 149, pad: 1 }), None)`

I don't know if that is being sent from wezterm or from X itself. It looks like rust dbug format though.

**Comment by tpoliaw** [comment]
I have rolled back to version 20220408-101518-b908e2dd and it is no longer happening. Also tried running the latest version of master (7adb5ac) and it is still happening there. If I can help in any way to debug this or if you'd more details on the environment it's running in, let me know.

**Comment by wez** [comment]
@tpoliaw please open a separate issue, and include the logs! 
You can find a log file in a location like `$XDG_RUNTIME_DIR/wezterm/wezterm-gui-log-PID.txt`

**Comment by wez** [comment]
@huibosa can you try the nightly build? We fixed #2198 and it may have been the same underlying issue as this one

**Comment by huibosa** [comment]
> @huibosa can you try the nightly build? We fixed #2198 and it may have been the same underlying issue as this one

Thank you for your work! I've test with the nightly build, and the bug is all fixed now
